### PR TITLE
Short-circuit for strings in WebUtility.UrlEncode

### PR DESCRIPTION
**Pull request by jamesqo**
This is a follow-up on #6543 for `UrlEncode`. It has separate implementations for making passes over the strings/byte arrays, so (I haven't tested this yet) there should be no performance loss.

cc @stephentoub

### CODE REVIEW

**Comment by davidsh** [comment]
>so (I haven't tested this yet) there should be no performance loss.

I would like to see performance data for this to show how this PR improves the current implementation.

**Comment by jamesqo** [comment]
@davidsh I seem to be getting a `FileNotFoundException` on loading an assembly when trying to profile the code from Visual Studio. I'd investigate this further, but I'll probably be busy for the next few weeks. It'd be awesome if someone else could take a look.

**Comment by jamesqo** [comment]
Never mind, I've sorted out the problems in VS. Here is the sample code I tested:

```csharp
static void Main()
{
    for (int i = 0; i < 10000; i++)
        WebUtility.UrlEncode("hello_world");
}
```

Before the changes:

before

After the changes (barely any allocations):

after

File: src/System.Runtime.Extensions/src/System/Net/WebUtility.cs [hunk 1]
```diff
@@ -351,19 +315,63 @@ private static byte[] UrlEncode(byte[] bytes, int offset, int count)
         #region UrlEncode public methods
 
         [SuppressMessage("Microsoft.Design", "CA1055:UriReturnValuesShouldNotBeStrings", Justification = "Already shipped public API; code moved here as part of API consolidation")]
-        public static string UrlEncode(string value)
+        public static unsafe string UrlEncode(string value)
         {
             if (string.IsNullOrEmpty(value))
```
Comment (davidsh) [inline src/System.Runtime.Extensions/src/System/Net/WebUtility.cs]: pls put braces around `if` clause.

File: src/System.Runtime.Extensions/src/System/Net/WebUtility.cs [hunk 2]
```diff
@@ -351,19 +315,63 @@ private static byte[] UrlEncode(byte[] bytes, int offset, int count)
...
+            bool foundSpaces = false;
+            int unsafeCount = 0;
+
+            // count them first
+            var encoding = Encoding.UTF8;
+            for (int i = 0; i < value.Length; i++)
+            {
+                char ch = value[i];
+
+                if (ch == ' ')
```
Comment (davidsh) [inline src/System.Runtime.Extensions/src/System/Net/WebUtility.cs]: pls put braces around `if` clause.

File: src/System.Runtime.Extensions/src/System/Net/WebUtility.cs [hunk 3]
```diff
@@ -351,19 +315,63 @@ private static byte[] UrlEncode(byte[] bytes, int offset, int count)
...
+                        1 : // UTF-8 uses 1 byte for chars 0 to 127
+                        encoding.GetByteCount(&ch, 1);
+                }
+            }
 
-            byte[] bytes = Encoding.UTF8.GetBytes(value);
-            byte[] encodedBytes = UrlEncode(bytes, 0, bytes.Length, false /* alwaysCreateNewReturnValue */);
-            return Encoding.UTF8.GetString(encodedBytes, 0, encodedBytes.Length);
+            // nothing to expand?
+            if (!foundSpaces && unsafeCount == 0)
```
Comment (davidsh) [inline src/System.Runtime.Extensions/src/System/Net/WebUtility.cs]: pls put braces around `if` clause.

**Comment by davidsh** [comment]
@jamesqo Thx for the perf data!

@stephentoub PTAL

**Comment by stephentoub** [comment]
Before your change, this code:
```
char[] chars = Enumerable.Range(0, ushort.MaxValue).Select(i => (char)i).ToArray();
string s = new string(chars);
Console.WriteLine(WebUtility.UrlEncode(s).Length);
```
prints out ```583141```.  After your change, it prints out ```583145```.

**Comment by jamesqo** [comment]
That's weird, it seems to be appending a bunch of nulls to the output. I narrowed down the problem to this code snippet:

```csharp
char[] chars = Enumerable.Range(0, ushort.MaxValue).Select(i => (char)i).ToArray();
string text = new string(chars);
var encoding = Encoding.UTF8;

Console.WriteLine(encoding.GetBytes(text).Length);
Console.WriteLine(text.Sum(ch => encoding.GetByteCount(&ch, 1)));
```

It prints two different numbers: `194427` for the whole string, and `194429` for counting each individually. I'll look into this further when I get back.

**Comment by jamesqo** [comment]
@stephentoub I've fixed the issue you've pointed out. Should I also add a test to prevent this regression in the future?

**Comment by davidsh** [comment]
>Should I also add a test to prevent this regression in the future?

Yes, add a new test.

**Comment by jamesqo** [comment]
@davidsh Thanks for the quick feedback! I've added the tests.

**Comment by JonHanna** [comment]
I think one should always add regression tests for anything one introduced out of pride if nothing else; "hey, if a smart person like me can do it, then surely someone else could" :smile:

**Comment by jamesqo** [comment]
@stephentoub Sorry, but could the *NO MERGE* label please be removed for this PR? I've fixed the issue you mentioned and tested the changes locally, so I think it's safe to be merged after code review. (For some reason, it doesn't seem to be running in CI, so I'll close/reopen this pull request to see if that will change things.)

**Comment by JonHanna** [comment]
Test  Innerloop Ubuntu Debug Build and Test 
Test Innerloop OSX Release Build and Test

File: src/System.Runtime.Extensions/src/System/Net/WebUtility.cs [hunk 4]
```diff
@@ -351,19 +315,82 @@ private static byte[] UrlEncode(byte[] bytes, int offset, int count)
...
+            for (int i = 0; i < count; i++)
+            {
+                char ch = (char)value[offset + i];
+
+                if (ch != ' ' && !IsUrlSafeChar(ch))
+                {
+                    unsafeCount++;
+                }
+            }
+
```
Comment (stephentoub) [inline src/System.Runtime.Extensions/src/System/Net/WebUtility.cs]: Why did you remove the checks that allow it to exit early without doing any encoding if no spaces and no unsafe chars were found?

Comment (jamesqo) [inline src/System.Runtime.Extensions/src/System/Net/WebUtility.cs]: You're right, I was only thinking about improvements allocation-wise, so I didn't consider the fact we would need need to make another pass over the array. I'll fix this in a moment.

**Comment by stephentoub** [comment]
@dotnet-bot test this please

File: src/System.Runtime.Extensions/src/System/Net/WebUtility.cs [hunk 5]
```diff
@@ -351,19 +315,90 @@ private static byte[] UrlEncode(byte[] bytes, int offset, int count)
...
+                {
+                    unsafeCount++;
+                }
+            }
+            
+            if (unsafeCount == 0)
+            {
+                // nothing to encode, just make a copy of the array
+                byte[] result = new byte[count];
+                Array.Copy(value, offset, result, 0, count);
```
Comment (jamesqo) [inline src/System.Runtime.Extensions/src/System/Net/WebUtility.cs]: Note: I just realized this change solves a pretty big issue where a copy of the entire array will be returned if no encoding is needed, since the old implementation uses `Array.Clone` to clone the entire array.

**Comment by jamesqo** [comment]
test innerloop osx release build and test
test innerloop ubuntu release build and test

**Comment by stephentoub** [comment]
@jamesqo, with your latest changes, this code:
```C#
byte[] input = Encoding.UTF8.GetBytes("this is a test");
byte[] output = WebUtility.UrlEncodeToBytes(input, 0, input.Length);
Console.WriteLine(output.Sum(i => (int)i)));
```
now prints out 1334 where it used to print out 1301.

I appreciate your efforts to try to make corefx better,  In this case, though, we've had a bunch of back and forths on this PR, and at this point I'm not confident enough in the level of testing being done to feel comfortable merging this.  It's not worth the risk to save on some allocations that haven't shown to be a problem.  @davidsh, up to you how you want to proceed.

**Comment by davidsh** [comment]
I agree with the uncertainty regarding test coverage.  I would like this PR to be closed. However, the added tests in this PR as useful and so feel free to submit a new PR with just added tests.  

And each new "failure" that was discovered while reviewing this PR should be incorporated into a new test:

i.e.

```c#
\byte[] input = Encoding.UTF8.GetBytes("this is a test");
byte[] output = WebUtility.UrlEncodeToBytes(input, 0, input.Length);
Console.WriteLine(output.Sum(i => (int)i)));
```
should be transformed or blended into the tests.

Once we have sufficient tests, then perhaps we can be more confident to think about refactoring for perf allocations.

**Comment by jamesqo** [comment]
@davidsh Sounds reasonable to me. I'll submit a new PR with extensive tests, and once that is merged (and after I've carefully tested this locally) I will re-submit this in a few weeks. Thanks!

File: src/System.Runtime.Extensions/src/System/Net/WebUtility.cs [hunk 6]
```diff
@@ -351,19 +315,90 @@ private static byte[] UrlEncode(byte[] bytes, int offset, int count)
...
+            }
+            
+            int unsafeCount = 0;
+
+            // count them first
+            for (int i = 0; i < count; i++)
+            {
+                char ch = (char)value[offset + i];
+
+                if (ch != ' ' && !IsUrlSafeChar(ch))
```
Comment (jamesqo) [inline src/System.Runtime.Extensions/src/System/Net/WebUtility.cs]: Looks like the problem was here. It should have looked like this instead:

```csharp
bool foundSpaces = false;
int unsafeCount = 0;

for (int i = 0; i < count; i++)
{
    char ch = (char)value[offset + i];

    if (ch == ' ')
    {
        foundSpaces = true;
    }
    else if (!IsUrlSafeChar(ch))
    {
        unsafeCount++;
    }
}

if (!foundSpaces && unsafeCount == 0)
{
    // ...
}
```

File: src/System.Runtime.Extensions/src/System/Net/WebUtility.cs [hunk 7]
```diff
@@ -351,19 +315,90 @@ private static byte[] UrlEncode(byte[] bytes, int offset, int count)
...
+                {
+                    if (ch < 128) // optimize for ASCII chars
+                    {
+                        unsafeCount++; // UTF-8 uses 1 byte for chars 0 to 127
+                    }
+                    else if (char.IsSurrogatePair(value, i)) // this method should handle the case where i is the last index
+                    {
+                        // surrogate code points are encoded differently separate than when alone, so we need to address that
+                        i++;
+                        unsafeCount += encoding.GetByteCount(&ch, 2);
```
Comment (jamesqo) [inline src/System.Runtime.Extensions/src/System/Net/WebUtility.cs]: Small optimization: this could be hard-coded as `4` instead (of course, it would have to be commented). See: http://ideone.com/vkpG92

**Comment by jamesqo** [comment]
@stephentoub Turns out it was actually a business logic-related error, rather than an obscure encoding error like the last one. When I responded to your PR feedback about making another pass over the array, I forgot to keep track of whether any spaces were found, so they simply weren't getting encoded ([relevant](https://github.com/dotnet/corefx/pull/6698/files#r56702096) line note). Nevertheless, I'll still hold off on re-submitting this until it gets more test coverage.

File: src/System.Runtime.Extensions/src/System/Net/WebUtility.cs (hunk 5 above)
Comment (jamesqo) [inline src/System.Runtime.Extensions/src/System/Net/WebUtility.cs]: Also: could be `Buffer.BlockCopy` instead.
